cpp/KernelData.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (C) 2022 Sarah Roggendorf
+//
+// This file is part of DOLFINx_Contact
+//
+// SPDX-License-Identifier:    MIT
+
+#include "KernelData.h"
+
+dolfinx_contact::KernelData::KernelData(
+    std::shared_ptr<const dolfinx::fem::FunctionSpace> V,
+    std::shared_ptr<const dolfinx_contact::QuadratureRule> q_rule,
+    const std::vector<std::size_t>& cstrides)
+    : _qp_offsets(q_rule->offset()), _q_weights(q_rule->weights())
+{
+  // Extract mesh
+  std::shared_ptr<const dolfinx::mesh::Mesh> mesh = V->mesh();
+  assert(mesh);
+  // Get mesh info
+  const dolfinx::mesh::Geometry& geometry = mesh->geometry();
+  const dolfinx::fem::CoordinateElement& cmap = geometry.cmap();
+  _affine = cmap.is_affine();
+  _num_coordinate_dofs = cmap.dim();
+
+  _gdim = geometry.dim();
+  const dolfinx::mesh::Topology& topology = mesh->topology();
+  _tdim = topology.dim();
+
+  if (const dolfinx::mesh::CellType ct = topology.cell_type();
+      (ct == dolfinx::mesh::CellType::prism)
+      || (ct == dolfinx::mesh::CellType::pyramid))
+  {
+    throw std::invalid_argument("Unsupported cell type");
+  }
+
+  // Create quadrature points on reference facet
+  const xt::xtensor<double, 2>& q_points = q_rule->points();
+  const std::size_t num_quadrature_pts = _q_weights.size();
+
+  // Extract function space data (assuming same test and trial space)
+  std::shared_ptr<const dolfinx::fem::FiniteElement> element = V->element();
+  std::shared_ptr<const dolfinx::fem::DofMap> dofmap = V->dofmap();
+  _ndofs_cell = dofmap->element_dof_layout().num_dofs();
+  _bs = dofmap->bs();
+
+  if (const bool needs_dof_transformations
+      = element->needs_dof_transformations();
+      needs_dof_transformations)
+  {
+    throw std::invalid_argument("Contact-kernels are not supporting finite "
+                                "elements requiring dof transformations.");
+  }
+
+  if (element->value_size() / _bs != 1)
+  {
+    throw std::invalid_argument(
+        "Contact kernel not supported for spaces with value size!=1");
+  }
+
+  if (_bs != _gdim)
+  {
+    throw std::invalid_argument(
+        "The geometric dimension of the mesh is not equal to the block size "
+        "of the function space.");
+  }
+
+  /// Pack test and trial functions
+  _phi = xt::xtensor<double, 2>({num_quadrature_pts, _ndofs_cell});
+  _dphi = xt::xtensor<double, 3>({_tdim, num_quadrature_pts, _ndofs_cell});
+  xt::xtensor<double, 4> basis_functions(
+      {_tdim + 1, num_quadrature_pts, _ndofs_cell, 1});
+  element->tabulate(basis_functions, q_points, 1);
```

Comment:
This can be simplified now that we can access the `basix::FiniteElement` from `dolfinx::fem::FiniteElement`, ref https://github.com/FEniCS/dolfinx/pull/2205